Source/WebCore/css/StyleProperties.cpp

```diff
@@ -576,12 +577,34 @@ String StyleProperties::fontValue() const
     if (auto shorthand = isSingleFontShorthand())
         return getValueNameAtomString(shorthand.value());
 
+    // Font stretch values can only be serialized as keywords.
+    // While it's not clear the CSS font specification calls for this, web platform tests expect us to convert percentages to keywords.
+    CSSValueID fontStretchPercentageAsKeyword = CSSValueInvalid;
+    if (int fontStretchPropertyIndex = findPropertyIndex(CSSPropertyFontStretch); fontStretchPropertyIndex != -1) {
+        if (auto value = dynamicDowncast<CSSPrimitiveValue>(*propertyAt(fontStretchPropertyIndex).value()); value && value->isPercentage()) {
+            auto valueAsDouble = value->doubleValue();
+            FontSelectionValue valueAsFontSelectionValue { static_cast<float>(valueAsDouble) };
+            auto valueAfterRoundTrip = valueAsFontSelectionValue;
+            if (valueAsDouble != valueAfterRoundTrip)
+                return emptyString();
+            auto keyword = fontStretchKeyword(valueAsFontSelectionValue);
+            if (!keyword)
+                return emptyString();
+            fontStretchPercentageAsKeyword = *keyword;
+        }
+    }
+
     String commonValue = fontSizeProperty.value()->cssText();
     StringBuilder result;
     appendFontLonghandValueIfExplicit(CSSPropertyFontStyle, result, commonValue);
     appendFontLonghandValueIfExplicit(CSSPropertyFontVariantCaps, result, commonValue);
     appendFontLonghandValueIfExplicit(CSSPropertyFontWeight, result, commonValue);
-    appendFontLonghandValueIfExplicit(CSSPropertyFontStretch, result, commonValue);
+    if (fontStretchPercentageAsKeyword != CSSValueInvalid) {
+        if (fontStretchPercentageAsKeyword != CSSValueNormal)
+            result.append(result.isEmpty() ? "" : " ", getValueName(fontStretchPercentageAsKeyword));
+        commonValue = { };
+    } else
+        appendFontLonghandValueIfExplicit(CSSPropertyFontStretch, result, commonValue);
```

Comment:
Took me a bit to get that this is hit in the case of a keyword being set explicitly.

I think we should also omit `CSSValueNormal` in this case too if we can.

This is in-line with the CSSOM spec: https://drafts.csswg.org/cssom/#serializing-css-values

> If component values can be omitted or replaced with a shorter representation without changing the meaning of the value, omit/replace them.

Reply:
Yes, good point. I think I need add a test case for that, because while you caught it the web tests did not! Which is annoying since I always find it slightly confusing and inconvenient to upstream WPT improvements. Maybe you can help me with that part?

The other reason I left that code around is that it handles the “common string” case which might be needed to do the right thing for special values like “inherit” and I might break them if I tried to stop using the helper function. So I decided to limit the change to cases I knew didn’t run into that.

Reply:
> I think I need add a test case for that, because while you caught it the web tests did not! Which is annoying since I always find it slightly confusing and inconvenient to upstream WPT improvements. Maybe you can help me with that part?

Yes! You can probably just extend the current test then use: `export-w3c-test-changes -g HEAD -b 246573 -c --no-linter` (given HEAD is this WebKit commit). It should just be 2 extra lines in the test:

```
    div.style.fontStretch = "normal";
    assert_equals(div.style.font, "medium serif", "The keyword normal should be omitted");
```

You can also make the PR manually from the web UI too: https://github.com/web-platform-tests/wpt/blob/master/css/css-fonts/font-shorthand-serialization-font-stretch.html

I can help with the export part if you don't have time, I'm ok with just making the change in the PR directly in our in-tree files for now.

Reply:
I tried the export but got stuck wondering where I put my OAuth token; I know that seems like a trivial reason to fail a this!

Reply:
https://github.com/darinadler/wpt/tree/wpt-export-for-webkit-246573

Reply:
@darinadler Looks great thanks! can you open a PR on the repo by clicking this link: https://github.com/darinadler/wpt/compare/web-platform-tests:wpt:master...wpt-export-for-webkit-246573?expand=1 and "Create pull request"? This way I can approve it and merge it.

This is basically the step that requires the OAuth token when done from the script, but it can be done manually from the web UI.